File: matlab code/runmsmDAHD4_L96.m
%clear all;
%close all;

load L96_F6.mat;
K=10;
data = center(data(1:K:end,:));
X=data(:,1:end);
W=160;
D=size(X,2);
NE=size(X,1);
NFE=W;
NP=D;
tic
wt='none';
wt='bartlett';





tic

[fE2,VP,FEP]=DAHD4freq_part_weight(X,W,NFE,NP,wt);



%%
NCMP=0.75*D; 
NCMP=20; 

NFEE=25;%% number of frequencies to emulate
NFEE=30;%% number of frequencies to emulate

ff =[2,10,20,30];
figure
semilogy(fE2,abs(VP),'ro','MarkerSize',4,'MarkerFaceColor','r');
hold on
semilogy(fE2(2:NFEE),abs(VP(1:NCMP,2:NFEE)),'bo','MarkerSize',4,'MarkerFaceColor','b');
semilogy(fE2(ff(1)),abs(VP(1,ff(1))),'ko','MarkerSize',8,'MarkerFaceColor','k');
semilogy(fE2(ff(2)),abs(VP(1,ff(2))),'ko','MarkerSize',8,'MarkerFaceColor','k');
semilogy(fE2(ff(3)),abs(VP(1,ff(3))),'ko','MarkerSize',8,'MarkerFaceColor','k');
semilogy(fE2(ff(4)),abs(VP(1,ff(4))),'ko','MarkerSize',8,'MarkerFaceColor','k');
ylabel('\lambda');
set(gca,'FontSize',16)
%%

NP=D;
EP=zeros((2*W-1)*D,2*NP,NFE);%% 
tic
for iff=1:NFE
ER=DAHM4_ex(FEP(:,:,iff),W,iff,2*NP);
ERR=reshape(ER,(2*W-1)*D,2*NP);
EP(:,:,iff)=ERR;
end
toc
%%

%%%%%%%%%%%%%%
fig=figure('Units','centimeter',...
		   'Position',[0.1 0.1 30 40],...
		   'DefaultAxesFontWeight', 'bold',...
		   'DefaultAxesFontSize', 10,'PaperUnits','centimeter','PaperPositionMode', 'auto','PaperSize',[30 40]);
kk=0;
for pos=1:4
kk=(pos-1)*3;
for K=1:2
kk=kk+1;
subplot(4,3,kk)
set(gca,'FontSize',16);
contourf(reshape(EP(:,K,ff(pos)),(2*W-1),D)','EdgeColor','none');
shading flat
colorbar
if K== 1 title(['f=' num2str(fE2(ff(pos)))]); end;
colormap('jet')
xlabel('Time')
ylabel('Space')
set(gca,'FontSize',16);
end
end

for pos=1:4
ER = reshape(EP(:,1:2,ff(pos)),size(EP,1),2);
A=dahc(X,ER);
subplot(4,3,pos*3)
plot(A,'LineWidth',2)
xlim([1 300])
xlabel('Time')
title(['f=' num2str(fE2(ff(pos)))]);
set(gca,'FontSize',16)
end



%%


%%%%%%%%%%%%%%%
NFS=1;
NFE=NFEE;
%%%%%%%%%%%%%%%
NT0=2;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
NEX = size(X,1);
RXT = zeros(NEX,size(X,2),NFE,NT0);
RRT = zeros(NEX,size(X,2),NFE);
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
WW=2*W-1;
NA=size(X,1)-WW+1;
NE = NEX-WW+1;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


XT_RES=zeros(NA,2*size(X,2));
for NF=NFS:NFE
%for NF=NFS:NFE
fprintf('FREQ=%d \n',NF);
%%%%%%%%% for W=20 %%%%%%%%%%%%%%%%%%%%%
if NF==1; %%
NM=size(X,2); %% for NF ==1;
ncmp = NM/2;
%ncmp = NM; 
end
if NF> 1
NM=(size(X,2))*2;
ncmp = NM;
ncmp = NM/2;
ncmp = 3*NM/4;
ncmp=2*NCMP;
end
%%%%%%%% for W=4 or W=10%%%%%%%%%%%%%%%%%%%%%
ER = reshape(EP(:,1:NM,NF),size(EP,1),NM);
A=dahc(X,ER);
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
indm=1:ncmp;
RR = hrc(A,ER,size(X,2),indm);
RRT(:,:,NF)=RR;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
RXZ = zeros(size(X,1),size(X,2),NT0);
         NN = size(A,1);
         %NE=4*NA;
         NE=NA;
         DMD=A(:,indm);
         ipls = 0;
         inorm =1;

         % ires=0;
         % L=11;

         ires=0;
         ires=0;
         L=5;
        
         rng('default');
         if ires == 0 indrandperm=randn(size(X,2)*2,NE,NT0);
         else indrandperm=zeros(NE-1,NT0);
         for NT=1:NT0
         indrandperm(:,NT)=randperm(NE-1);
         end
         end
         %NT=NT0;
         %XT_RES=zeros(NA,2*size(X,2));
if NF==1
[xx,xt_res,LL,ENL]= MSLM_INT(NE,DMD(:,:),0,L,NT0,ires,inorm,0,0,0,indrandperm);%%
%XT_RES(1:NA-L,1:ncmp)=xt_res;
else
iNL=1;
iSYM = 1;
inEQ = 1;
[xx,xt_res,LL,ENL]= MSLM_INT(NE,DMD(:,:),1,L,NT0,ires,inorm,iSYM,iNL,inEQ,indrandperm);%%
%display([num2str(ENL','%.5f')])
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%XT_RES(1:NA-L,1:ncmp)=xt_res;
end

for KK=1:NT0
indf = indm;
RXZ(:,:,KK) = hrc(xx(:,indf,KK),ER(:,indf),size(X,2),1:size(A(:,indf),2));
end

%RXT(:,:,NF-NFS+1,:)=RXZ; %no parfor
%RES(:,:,NF-NFS+1)=XT_RES;

RXT(:,:,NF,:)=RXZ; %% parfor
%RES(:,:,NF)=XT_RES;


        end
%%
     
                 NET=2000;
                 figure
                 for KK=1:size(RXT,4)
                 
                 indf = 1:NFE-NFS+1;
              
                 RR=sum(RRT(:,:,:),3);
                 
                 RX=squeeze(sum(RXT(:,:,:,KK),3));
                 display(['Simulation #' num2str(KK) ' Var Recons: ' num2str(sum(var(RR))) ' Var MSLM: '  num2str(sum(var(RX))) ])
                 
                 ND=1;
                indt=NEX-NET:NEX;
                 cmax = 8;
                 cmin = -8;
                 subplot(313)
                contourf(RX(indt,:)',20,'EdgeColor','none')
                         
                          shading flat
                          colorbar
                          colormap('jet')
                          caxis([cmin cmax]);
                          title('MSLM')
                           xlabel('Time')
                           ylabel('Space')
                            set(gca,'FontSize',16)
                          subplot(312)
                         contourf(RR(indt,:)',20,'EdgeColor','none')
                      
                                   shading flat
                                   colorbar
                                   colormap('jet')
                                   caxis([cmin cmax]);
                                   title('DAHD Reconstruction')
                                   xlabel('Time')
                                   ylabel('Space')
                                     set(gca,'FontSize',16)

                                      subplot(311)
                         contourf(X(indt,:)',20,'EdgeColor','none')
                      
                                   shading flat
                                   colorbar
                                   colormap('jet')
                                   caxis([cmin cmax]);
                                   title('Full Data')
                                   xlabel('Time')
                                   ylabel('Space')
                                     set(gca,'FontSize',16)
                                           %pause
                                   end
                                   
                                   
                                   figure
                                 
                                   plot(squeeze(sum(var(RRT))),'r','LineWidth',2)
                                   hold on
                                   plot(squeeze(sum(var(RXT))),'b','LineWidth',2)
                                   legend('Recons','MSLM','Location','NorthWest')
                                   xlabel('Freq #')
                                   ylabel('variance')
                                   set(gca,'FontSize',16)
                                   
                                  
                                   
                                
                                
       statglob(sum(RRT(:,:,:),3),squeeze(sum(RXT(:,:,:,:),3)),300);
                                   
                                   
          
                                  
                                  
                                  
                                  
                                    
                                   
                                  
                           

File: runmsmDAHD4_L96.py
"""
runmsmDAHD4_L96.py

Top-level driver: DAHD spectral analysis + multi-level stochastic
Stuart-Landau (MSLM) simulation of the Lorenz-96 model (L96_F6.mat).

Python conversion of runmsmDAHD4_L96.m (Dmitri Kondrashov) by
Taylor McDonald and Dmitri Kondrashov 2026,  reusing the shared DAHD4 package
(center, dahc, hrc, DAHD4freq_part_weight, DAHM4_ex) plus the new
MSLM_INT.py / statglob.py translated alongside this script.

The original MATLAB used `parfor` for the per-frequency MSLM loop; this
translation runs it serially (the iterations are independent, so it can be
parallelized with e.g. `joblib.Parallel` if desired -- not done here to keep
the translation straightforward).
"""

import os
import sys

import numpy as np
import matplotlib.pyplot as plt
import scipy.io

# ── Make the DAHD4 package importable ───────────────────────────────────────
_DAHD4_DIR = os.environ.get("DAHD4_DIR", "/Users/dmitri/PycharmProjects/DAHD4")
if _DAHD4_DIR not in sys.path:
    sys.path.insert(0, _DAHD4_DIR)

from center import center  # noqa: E402
from dahc import dahc  # noqa: E402
from hrc import hrc  # noqa: E402
from DAHD4freq_part_weight import DAHD4freq_part_weight  # noqa: E402
from DAHM4_ex import DAHM4_ex  # noqa: E402

from MSLM_INT import MSLM_INT
from statglob import statglob


# ── 1. Load & prepare L96 data ──────────────────────────────────────────────
mat = scipy.io.loadmat('L96_F6.mat')
data_raw = mat['data']

K = 10
data_ds = data_raw[::K, :]
data_c, _ = center(data_ds)
X = data_c

W = 160
D = X.shape[1]
NE = X.shape[0]
NFE = W
NP = D
wt = 'bartlett'

print("Running DAHD spectral analysis...")
fE2, VP, FEP = DAHD4freq_part_weight(X, W, NFE, NP, wt)

# ── 2. Spectrum plot ─────────────────────────────────────────────────────────
# NCMP =  int(round(0.5 * D))
NCMP = 20
NFEE = 30  # number of frequencies to emulate
ff = [1, 9, 19, 29]  # (MATLAB ff = [2,10,20,30], 1-indexed)

plt.figure()
plt.semilogy(fE2, np.abs(VP).T, 'ro', markersize=4, markerfacecolor='r')
plt.semilogy(fE2[1:NFEE], np.abs(VP[:int(NCMP), 1:NFEE]).T, 'bo',
             markersize=4, markerfacecolor='b')
for pos in ff:
    plt.semilogy(fE2[pos], np.abs(VP[0, pos]), 'ko', markersize=8, markerfacecolor='k')
plt.ylabel(r'$\lambda$')
plt.tick_params(labelsize=16)
plt.tight_layout()
plt.show()

# ── 3. Compute DAHMs (space-time modes) ─────────────────────────────────────
NP = D
EP = np.zeros(((2 * W - 1) * D, 2 * NP, NFE))
print("Computing DAHMs...")
for iff in range(NFE):
    ER = DAHM4_ex(FEP[:, :, iff], W, iff, 2 * NP)
    EP[:, :, iff] = np.reshape(ER, ((2 * W - 1) * D, 2 * NP))

# ── 4. Mode contour plots + DAHC coefficient time series (4 selected freqs) ─
fig, axes = plt.subplots(4, 3, figsize=(15, 18))
for pos in range(4):
    for Kc in range(2):
        ax = axes[pos, Kc]
        cf = ax.contourf(EP[:, Kc, ff[pos]].reshape(2 * W - 1, D).T, 20, cmap='jet')
        plt.colorbar(cf, ax=ax)
        if Kc == 0:
            ax.set_title(f'f={fE2[ff[pos]]:.4g}')
        ax.set_xlabel('Time')
        ax.set_ylabel('Space')
        ax.tick_params(labelsize=16)

for pos in range(4):
    ER4 = EP[:, 0:2, ff[pos]]
    A = dahc(X, ER4)
    ax = axes[pos, 2]
    ax.plot(A, linewidth=2)
    ax.set_xlim([0, 300])
    ax.set_xlabel('Time')
    ax.set_title(f'f={fE2[ff[pos]]:.4g}')
    ax.tick_params(labelsize=16)
plt.tight_layout()
plt.show()

# ── 5. MSLM simulation, per frequency band ──────────────────────────────────
NFS = 0
NFE = NFEE
NT0 = 2

NEX = X.shape[0]
RXT = np.zeros((NEX, D, NFE, NT0))
RRT = np.zeros((NEX, D, NFE))

WW = 2 * W - 1
NA = X.shape[0] - WW + 1
NE = NEX - WW + 1

print("Running MSLM simulation per frequency band...")
for NF in range(NFS, NFE):
    print(f'FREQ={NF + 1}')

    if NF == 0:
        NM = D
        ncmp = D // 2
    else:
        NM = D * 2
        # (original MATLAB reassigns ncmp several times; only the final
        # value -- ncmp = 2*NCMP -- is actually used)
        ncmp = int(round(2 * NCMP))

    ER = EP[:, :NM, NF]
    A = dahc(X, ER)

    indm = np.arange(ncmp)
    RR = hrc(A, ER, D, indm)
    RRT[:, :, NF] = RR

    DMD = A[:, indm]
    inorm = 1
    # (original MATLAB sets ires=0 then immediately overwrites it with
    # ires=1; only the final value takes effect)
    ires = 0
    L_lvl = 5

    np.random.seed(0)  # rng('default')
    if ires == 0:
        indrandperm = np.random.randn(D * 2, NE, NT0)
    else:
        indrandperm = np.zeros((NE - 1, NT0), dtype=int)
        for NT in range(NT0):
            indrandperm[:, NT] = np.random.permutation(NE - 1)

    if NF == 0:
        xx, xt_res, LL, ENL = MSLM_INT(NE, DMD, 0, L_lvl, NT0, ires, inorm,
                                        0, 0, 0, indrandperm)
    else:
        iNL, iSYM, inEQ = 1, 1, 1
        xx, xt_res, LL, ENL = MSLM_INT(NE, DMD, 1, L_lvl, NT0, ires, inorm,
                                        iSYM, iNL, inEQ, indrandperm)
        #print(ENL.T)
    RXZ = np.zeros((NEX, D, NT0))
    for KK in range(NT0):
        indf = indm
        RXZ[:, :, KK] = hrc(xx[:, indf, KK], ER[:, indf], D, np.arange(len(indf)))

    RXT[:, :, NF, :] = RXZ

# ── 6. Reconstruction vs. MSLM comparison plots ─────────────────────────────
NET = 2000
RR = np.sum(RRT, axis=2)

#fig, (ax_top, ax_bot) = plt.subplots(3, 1, figsize=(10, 12))
fig, axs = plt.subplots(3, 1, figsize=(10, 12))  # Adjust figsize as needed
for KK in range(RXT.shape[3]):
    RX = np.sum(RXT[:, :, :, KK], axis=2)
    print(f"Simulation #{KK + 1} "
          f"Var Recons: {np.sum(np.var(RR, axis=0, ddof=1)):.6g} "
          f"Var MSLM: {np.sum(np.var(RX, axis=0, ddof=1)):.6g}")

    indt = np.arange(max(NEX - NET - 1, 0), NEX)
    cmin, cmax = -8, 8
    levels = np.linspace(cmin, cmax, 21)

    # cf1 = ax_bot.contourf(RX[indt, :].T, levels=levels, cmap='jet', extend='both')
    # plt.colorbar(cf1, ax=ax_bot)
    # ax_bot.set_title('MSLM'); ax_bot.set_xlabel('Time'); ax_bot.set_ylabel('Space')
    # ax_bot.tick_params(labelsize=16)
    #
    # cf2 = ax_top.contourf(RR[indt, :].T, levels=levels, cmap='jet', extend='both')
    # plt.colorbar(cf2, ax=ax_top)
    # ax_top.set_title('Reconstruction'); ax_top.set_xlabel('Time'); ax_top.set_ylabel('Space')
    # ax_top.tick_params(labelsize=16)
    # Subplot 311 - Full Data
    cf1 = axs[0].contourf(X[indt, :].T, levels=20, cmap='jet')
    axs[0].set_title('Full Data', fontsize=16)
    axs[0].set_xlabel('Time', fontsize=16)
    axs[0].set_ylabel('Space', fontsize=16)
    axs[0].tick_params(labelsize=16)
    fig.colorbar(cf1, ax=axs[0])
    cf1.set_clim(cmin, cmax)

    # Subplot 312 - DAHD Reconstruction
    cf2 = axs[1].contourf(RR[indt, :].T, levels=20, cmap='jet')
    axs[1].set_title('DAHD Reconstruction', fontsize=16)
    axs[1].set_xlabel('Time', fontsize=16)
    axs[1].set_ylabel('Space', fontsize=16)
    axs[1].tick_params(labelsize=16)
    fig.colorbar(cf2, ax=axs[1])
    cf2.set_clim(cmin, cmax)

    # Subplot 313 - MSLM
    cf3 = axs[2].contourf(RX[indt, :].T, levels=20, cmap='jet')
    axs[2].set_title('MSLM', fontsize=16)
    axs[2].set_xlabel('Time', fontsize=16)
    axs[2].set_ylabel('Space', fontsize=16)
    axs[2].tick_params(labelsize=16)
    fig.colorbar(cf3, ax=axs[2])
    cf3.set_clim(cmin, cmax)

plt.tight_layout()
plt.show()

if NFE - NFS > 1:
    var_RRT = np.sum(np.var(RRT, axis=0, ddof=1), axis=0)          # (NFE,)
    var_RXT = np.sum(np.var(RXT, axis=0, ddof=1), axis=0)          # (NFE, NT0)

    plt.figure()
    plt.plot(var_RRT, 'r', linewidth=2, label='Recons')
    for k in range(var_RXT.shape[1]):
        plt.plot(var_RXT[:, k], 'b', linewidth=2, label='MSLM' if k == 0 else None)
    plt.legend(loc='upper left')
    plt.xlabel('Freq #')
    plt.ylabel('variance')
    plt.tick_params(labelsize=16)
    plt.tight_layout()
    plt.show()

# ── 7. Autocorrelation / PDF diagnostics across the ensemble ────────────────
if RXT.shape[3] > 1:
    RX_all = np.sum(RXT, axis=2)                       # (NEX, D, NT0)
    datas_for_stat = RX_all.reshape(NEX, D * NT0)       # collapse to (N, K), matching
                                                         # MATLAB's implicit size() flattening
    statglob(RR, datas_for_stat, 300)
